tests/ui/collapsible_span_lint_calls.rs

```diff
@@ -0,0 +1,52 @@
+// run-rustfix
+#![deny(clippy::internal)]
+ #![feature(rustc_private)]
+
+ extern crate rustc_session;
+ extern crate rustc_errors;
+ extern crate rustc_lint;
+ extern crate syntax;
+ extern crate rustc_span;
+
+use rustc_session::{declare_tool_lint, declare_lint_pass};
+use rustc_errors::{DiagnosticBuilder, Applicability};
+use rustc_lint::{EarlyContext, EarlyLintPass, LintContext, Lint};
+use syntax::ast::Expr;
+use rustc_span::source_map::Span;
+
+#[allow(unused_variables)]
+pub fn span_lint_and_then<'a, T: LintContext, F>(cx: &'a T, lint: &'static Lint, sp: Span, msg: &str, f: F)
+where
+    F: for<'b> FnOnce(&mut DiagnosticBuilder<'b>),
+{
+}
+
+declare_tool_lint!{
+    pub clippy::TEST_LINT,
+    Warn,
+    "",
+    report_in_external_macro: true
+}
+
+declare_lint_pass!(Pass => [TEST_LINT]);
+
+impl EarlyLintPass for Pass {
+    fn check_expr(&mut self, cx: &EarlyContext, expr: &Expr)  {
+        let lint_msg = "lint message";
+        let help_msg = "help message";
+        let note_msg = "note message";
+        let sugg = "new_call()";
+
+        span_lint_and_then(cx, TEST_LINT, expr.span, lint_msg, |db| {
+            db.span_suggestion(expr.span, help_msg, sugg.to_string(), Applicability::MachineApplicable);
+        });
+        span_lint_and_then(cx, TEST_LINT, expr.span, lint_msg, |db| {
+            db.span_help(expr.span, help_msg);
+        });
+        span_lint_and_then(cx, TEST_LINT, expr.span, lint_msg, |db| {
+            db.span_note(expr.span, note_msg);
+        });
```

Comment:
Can you add a test where the spans are different? (Should not lint) Also a test, where something else is done in the closure would be good.

Reply:
How can I make up a different Span?

Reply:
`if let ExprKind::Unop(e) = expr.kind { e.span }`  pseudocode obviously

Reply:
Oh, suddenly realize rust won't emit warning on this snippet.

Reply:
With a different span, it should only warn on `and_then`+`span_note`/`and_then`+`note`